yarn-project/ethereum/src/l1_tx_utils/readonly_l1_tx_utils.ts

```diff
@@ -63,6 +63,119 @@ export class ReadOnlyL1TxUtils {
+   * Analyzes pending transactions and recent fee history to determine a competitive priority fee.
+   * Falls back to network estimate if data is unavailable or fails.
+   * @param networkEstimate - The base network priority fee estimate
+   * @returns A competitive priority fee based on pending txs and recent block history
+   */
+  protected async getCompetitivePriorityFee(networkEstimate: bigint): Promise<bigint> {
+    let competitiveFee = networkEstimate;
+
+    // 1. Check pending transactions (current competition)
+    try {
+      const pendingBlock = await this.client.getBlock({ blockTag: 'pending', includeTransactions: true });
+
+      if (pendingBlock.transactions && pendingBlock.transactions.length > 0) {
+        // Extract priority fees from pending transactions
+        const pendingFees = pendingBlock.transactions
+          .map((tx: any) => {
+            const fee = tx.maxPriorityFeePerGas || 0n;
+            // Debug: Log suspicious fees
+            if (fee > 100n * WEI_CONST) {
+              this.logger?.warn('Suspicious high priority fee in pending tx', {
+                txHash: tx.hash,
+                maxPriorityFeePerGas: formatGwei(fee),
+                maxFeePerGas: formatGwei(tx.maxFeePerGas || 0n),
+                maxFeePerBlobGas: tx.maxFeePerBlobGas ? formatGwei(tx.maxFeePerBlobGas) : 'N/A',
+              });
+            }
+            return fee;
+          })
+          .filter((fee: bigint) => fee > 0n);
+
+        if (pendingFees.length > 0) {
+          // Use 75th percentile of pending fees to be competitive
+          const sortedPendingFees = [...pendingFees].sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
+          const percentile75Index = Math.floor(sortedPendingFees.length * 0.75);
+          const pendingCompetitiveFee = sortedPendingFees[percentile75Index];
+
+          if (pendingCompetitiveFee > competitiveFee) {
+            competitiveFee = pendingCompetitiveFee;
+          }
+
+          this.logger?.debug('Analyzed pending transactions for competitive pricing', {
+            pendingTxCount: pendingFees.length,
+            pendingP75: formatGwei(pendingCompetitiveFee),
+          });
+        }
+      }
+    } catch (error) {
+      // Some nodes don't support pending block with transactions, that's ok
+      this.logger?.debug('Could not get pending block transactions', { error: (error as Error).message });
+    }
+
+    // 2. Check fee history (recent trends)
+    try {
+      const feeHistory = await this.client.getFeeHistory({
+        blockCount: 10,
+        rewardPercentiles: [25, 50, 75],
+      });
+
+      if (feeHistory.reward && feeHistory.reward.length > 0) {
+        // Extract 75th percentile fees from each block (index 2 in rewardPercentiles)
+        const percentile75Fees = feeHistory.reward.map(rewards => rewards[2] || 0n).filter(fee => fee > 0n);
+
+        if (percentile75Fees.length > 0) {
+          // Calculate median of the 75th percentile fees across blocks
```

Comment:
Curious why use the median of the last 10 blocks? I'm worried about a situation with increasing priority fees across blocks, and this would return a good priority fee for 5 blocks ago, not for the next one (assuming the fee estimate from the pending txs failed, that is).

Reply:
FWIW I checked what the last version of OZ relayers was doing, and they are checking the last 4 blocks: https://github.com/OpenZeppelin/openzeppelin-relayer/blob/471128d51951bf6cb89dcd206ce821bd5146937c/src/services/gas/evm_gas_price.rs#L368-L389

Not saying we change it to that, but wanted to make sure we're good with 10.